daemon/src/identify/dialer.rs

```diff
@@ -0,0 +1,166 @@
+use crate::identify::protocol;
+use crate::identify::PeerInfo;
+use crate::identify::PROTOCOL;
+use async_trait::async_trait;
+use std::collections::HashMap;
+use tokio::sync::watch;
+use tokio_extras::spawn_fallible;
+use xtra::Address;
+use xtra::Context;
+use xtra_libp2p::endpoint;
+use xtra_libp2p::libp2p::PeerId;
+use xtra_libp2p::Endpoint;
+use xtra_libp2p::OpenSubstream;
+use xtra_productivity::xtra_productivity;
+
+pub struct Actor {
+    endpoint: Address<Endpoint>,
+    peer_infos: HashMap<PeerId, PeerInfo>,
+    peer_info_channel: Option<watch::Sender<Option<PeerInfo>>>,
+}
+
+impl Actor {
+    pub fn new(endpoint: Address<Endpoint>) -> Self {
+        NUM_LIBP2P_CONNECTIONS_GAUGE.reset();
+
+        Self {
+            endpoint,
+            peer_infos: HashMap::default(),
+            peer_info_channel: None,
+        }
+    }
+
+    pub fn new_with_subscriber(
+        endpoint: Address<Endpoint>,
+    ) -> (Self, watch::Receiver<Option<PeerInfo>>) {
+        NUM_LIBP2P_CONNECTIONS_GAUGE.reset();
+
+        let (sender, receiver) = watch::channel::<Option<PeerInfo>>(None);
+
+        (
+            Self {
+                endpoint,
+                peer_infos: HashMap::default(),
+                peer_info_channel: Some(sender),
+            },
+            receiver,
+        )
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = ();
+
+    async fn stopped(self) -> Self::Stop {}
+}
+
+pub(crate) struct IdentifyMsgReceived {
+    peer_id: PeerId,
+    identify_msg: protocol::IdentifyMsg,
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(&mut self, msg: IdentifyMsgReceived) {
+        let peer_id = msg.peer_id;
+        let peer_info = match PeerInfo::try_from(msg.identify_msg.clone()) {
+            Ok(peer_info) => peer_info,
+            Err(e) => {
+                tracing::error!("Peer info discarded {:?}: {e:#}", msg.identify_msg);
+                return;
+            }
+        };
+
+        let wire_version = peer_info.wire_version.clone();
+        let daemon_version = peer_info.daemon_version.clone();
+        let environment = peer_info.environment;
+
+        NUM_LIBP2P_CONNECTIONS_GAUGE
+            .with(&HashMap::from([
+                (WIRE_VERSION_LABEL, wire_version.as_str()),
+                (DAEMON_VERSION_LABEL, daemon_version.as_str()),
+                (ENVIRONMENT_LABEL, environment.to_string().as_str()),
+            ]))
+            .inc();
+
+        tracing::info!(%peer_id, %daemon_version, %environment, %wire_version, "New identify message received");
+        self.peer_infos.insert(peer_id, peer_info.clone());
+
+        if let Some(peer_info_channel) = &self.peer_info_channel {
+            if let Err(e) = peer_info_channel.send(Some(peer_info)) {
+                tracing::warn!("Failed to send identity info to notify channel: {e:#}");
+            }
+        }
+    }
+}
+
+#[xtra_productivity(message_impl = false)]
```

Comment:
```suggestion
#[xtra_productivity]
```

message_impl does not exist anymore :)